### PR TITLE
Update Cryptographic Storage CS

### PR DESCRIPTION
**Pull request by rbsec**
Various updates to the structure and contents of the cryptographic storage CS, as per discussion in #324  and other known issues with the CS.

### CODE REVIEW

**Comment by rbsec** [comment]
Right, I think this is ready for an initial (content) review. I've tried to keep the recommendations more or less the same as the original, just with a lot of fluff and overly technical stuff cut out.

@ThunderSon @mackowski and anyone else - comments are very welcome.

File: cheatsheets/Cryptographic_Storage_Cheat_Sheet.md [hunk 1]
```diff
@@ -1,234 +1,175 @@
...
+Where possible, encryption keys should be stored in a separate location from encrypted data. For example, if the data is stored in a database, the keys should be stored in the filesystem. This means that if an attacker only has access to one of these (for example through directory traversal or SQL injection), they cannot access both the keys and the data.
 
-To perform the strong key management functions we have seen in requirement 3.6 we must have highly trusted and trained key custodians who understand how to perform key management duties. The key custodians must also sign a form stating they understand the responsibilities that come with this role.
+Depending on the architecture of the environment, it may be possible to store the keys and data on separate systems, which would provide a greater degree of isolation.
 
-# Related documentation & tools
+### Encrypting Stored Keys
 
-## Documentation
+Where possible, encryption keys should themselves be stored in an encrypted form. At least two separate keys are required for this:
```
Comment (mackowski) [inline cheatsheets/Cryptographic_Storage_Cheat_Sheet.md]: This is well described here: https://github.com/google/tink/blob/master/docs/JAVA-HOWTO.md#envelope-encryption.

We should mention why this is good approach (thread model)

File: cheatsheets/Cryptographic_Storage_Cheat_Sheet.md [hunk 2]
```diff
@@ -1,234 +1,175 @@
...
 
-###### 3.6.4 Periodic cryptographic key changes
+It is important that the code and processes required to rotate a key are in place **before** they are required, so that keys can be quickly rotated in the event of a compromise. Additionally, processes should also be implemented to allow the encryption algorithm or library to be changed, in case a new vulnerability is found in the algorithm or implementation.
 
-The PCI DSS standard mandates that keys used for encryption must be rotated at least annually. The key rotation process must remove an old key from the encryption/decryption process and replace it with a new key. All new data entering the system must encrypted with the new key. While it is recommended that existing data be rekeyed with the new key, as per the Rekey data at least every one to three years rule above, it is not clear that the PCI DSS requires this.
+## Key Storage
```
Comment (mackowski) [inline cheatsheets/Cryptographic_Storage_Cheat_Sheet.md]: It would be good to briefly describe why HSM/KMS itp is better than just encrypted file on disk

Comment (rbsec) [reply cheatsheets/Cryptographic_Storage_Cheat_Sheet.md]: I've added some notes about the benefits of KMS/HSM solutions.

File: cheatsheets/Cryptographic_Storage_Cheat_Sheet.md (hunk 1 above)
Comment (rbsec) [reply cheatsheets/Cryptographic_Storage_Cheat_Sheet.md]: I've expanded this a bit and linked to some Google documentation talking about DEK/KEK usage patterns.

File: cheatsheets/Cryptographic_Storage_Cheat_Sheet.md [hunk 3]
```diff
@@ -1,234 +1,185 @@
...
+The best way to protect sensitive information is to not store it in the first place. Although this applies to all kinds of information, it is most often applicable to credit card details, as they are highly desirable for attackers, and PCI DSS has such stringent requirements for how they must be stored. Wherever possible, the storage of sensitive information should be avoided.
 
-If a password is being used to protect keys then the [password strength](http://en.wikipedia.org/wiki/Password_strength) should be sufficient for the strength of the keys it is protecting.
+## Algorithms
 
-When 3DES is used, ensure `K1 != K2 != K3`, and the minimum key length must be `192 bits` .
+For symmetric encryption **AES** with a **256 bit** key should be used as the preferred algorithm.
 
-#### Rule - Use approved cryptographic modes
+For asymmetric encryption, **RSA** with at least a **2048 bit** key should be used as the preferred algorithm.
```
[the remark this answers is not in the document]
Comment (rbsec) [reply cheatsheets/Cryptographic_Storage_Cheat_Sheet.md]: It's a fair point - RSA is very widely supported at the moment, but if there's reasonable support in languages then I think you're right that we should be pointing to ECC crypto such as `X25519` (ed25519 is for signing, not encrypting).

[the remark this answers is not in the document]
Comment (rbsec) [reply cheatsheets/Cryptographic_Storage_Cheat_Sheet.md]: Updated in 86c86aa.

File: cheatsheets/Cryptographic_Storage_Cheat_Sheet.md [hunk 4]
```diff
@@ -1,234 +1,185 @@
...
-- In general, hash lengths are twice as long as the security bits offered by the symmetric/asymmetric algorithm  e.g. SHA-224 for 3TDEA (112 security bits) (due to the [Birthday Attack](http://en.wikipedia.org/wiki/Birthday_attack))
+The best way to protect sensitive information is to not store it in the first place. Although this applies to all kinds of information, it is most often applicable to credit card details, as they are highly desirable for attackers, and PCI DSS has such stringent requirements for how they must be stored. Wherever possible, the storage of sensitive information should be avoided.
 
-If a password is being used to protect keys then the [password strength](http://en.wikipedia.org/wiki/Password_strength) should be sufficient for the strength of the keys it is protecting.
+## Algorithms
 
-When 3DES is used, ensure `K1 != K2 != K3`, and the minimum key length must be `192 bits` .
+For symmetric encryption **AES** with a **256 bit** key should be used as the preferred algorithm.
```
[the remark this answers is not in the document]
Comment (rbsec) [reply cheatsheets/Cryptographic_Storage_Cheat_Sheet.md]: Much like TLSv1.1, while there's nothing wrong with AES 128, to my mind there's also not really any reason to use it for the vast majority of people. It also gives you a much bigger safety margin against future attacks and improvements.

I suppose we could add something basically saying "There's nothing really wrong with AES 128, but you should be using AES 256", but I'm wondering if that just confuses the issue a bit? @mackowski do you have any thoughts on this?

File: cheatsheets/Cryptographic_Storage_Cheat_Sheet.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (ThunderSon) [reply cheatsheets/Cryptographic_Storage_Cheat_Sheet.md]: @chick3nman what do you think could be a middle point for such a discussion? Between the point of RSA slowly going away, and not enough support for the new functions.

File: cheatsheets/Cryptographic_Storage_Cheat_Sheet.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (mackowski) [reply cheatsheets/Cryptographic_Storage_Cheat_Sheet.md]: What about:

> For symmetric encryption **AES** with a key **at least 128 bits** (preferable **256 bit**) should be used as the preferred algorithm.

**Comment by readmeexe** [comment]
Line 52 
```[PCI DSS](#34-render-pan-primary-account-number-at-minimum-unreadable-anywhere-it-is-stored) ```
refers to section that doesn't exist in this version.
Otherwise, very nice update.

**Comment by rbsec** [comment]
@readmeexe thanks for the review - really appreciated. I've fixed the PCI link in c572c18 to point to the PCI document library (don't want to link to an individual version of PCI DSS< because the link will just break when the next update comes out).

File: cheatsheets/Cryptographic_Storage_Cheat_Sheet.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (rbsec) [reply cheatsheets/Cryptographic_Storage_Cheat_Sheet.md]: Updated to

> For symmetric encryption **AES** with a key that's at least **128 bits** (ideally **256 bits**) and a secure [mode](#cipher-modes) should be used as the preferred algorithm.

File: cheatsheets/Cryptographic_Storage_Cheat_Sheet.md [hunk 5]
```diff
@@ -1,234 +1,185 @@
...
+The best way to protect sensitive information is to not store it in the first place. Although this applies to all kinds of information, it is most often applicable to credit card details, as they are highly desirable for attackers, and PCI DSS has such stringent requirements for how they must be stored. Wherever possible, the storage of sensitive information should be avoided.
 
-If a password is being used to protect keys then the [password strength](http://en.wikipedia.org/wiki/Password_strength) should be sufficient for the strength of the keys it is protecting.
+## Algorithms
 
-When 3DES is used, ensure `K1 != K2 != K3`, and the minimum key length must be `192 bits` .
+For symmetric encryption **AES** with a key that's at least **128 bits** (ideally **256 bits**) and a secure [mode](#cipher-modes) should be used as the preferred algorithm.
 
-#### Rule - Use approved cryptographic modes
+For asymmetric encryption, use elliptical curve cryptography (ECC) with a secure curve such as **X25519** as a preferred algorithm. If ECC is not available, **RSA** must be used with a key that is at least **2048 bits** in size.
```
Comment (Chick3nman) [inline cheatsheets/Cryptographic_Storage_Cheat_Sheet.md]: Change 'X25519' to 'Curve25519' to fit proper naming convention or change sentence structure. The Curve primitive is called 'Curve25519', the DH function 'X25519', and the EdDSA function 'Ed25519'.

File: cheatsheets/Cryptographic_Storage_Cheat_Sheet.md [hunk 6]
```diff
@@ -1,234 +1,185 @@
...
-##### 3.6 Fully document and implement all key-management processes and procedures for cryptographic keys used for encryption of cardholder data
+Encryption keys should be changed (or rotated) based on a number of different criteria:
 
-Requirement 3.6 mandates that key management processes within a PCI compliant company cover 8 specific key lifecycle steps:
+- If the previous key is known (or suspected) to have been compromised.
+  * This could also be caused by a someone who had access to the key leaving the organisation.
+- After a specified period of time has elapsed (known as the cryptoperiod).
+  * There are many factors that could affect what an appropriate cryptoperiod is, including the size of the key, the sensitivity of the data, and the threat model of the system. See section 5.3 of [NIST SP 800-57](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-57pt1r4.pdf) for further guidance.
+- After the key has been used to encrypt a specific amount of data.
+  * This would typically be `2^35` bytes (~34GB) for 64-bit keys (DES, 3DES, Blowfish, RC5, etc.) and `2^68` bytes (~295 exabytes) for 128 bit keys (AES, Twofish, Serpent, etc.).
```
Comment (Chick3nman) [inline cheatsheets/Cryptographic_Storage_Cheat_Sheet.md]: DES has a 64bit block size but not a 64bit key size, it has a 56bit key size. 3DES has either a 128bit or 192bit key size. This area needs review.

File: cheatsheets/Cryptographic_Storage_Cheat_Sheet.md (hunk 5 above)
Comment (rbsec) [reply cheatsheets/Cryptographic_Storage_Cheat_Sheet.md]: Fixed in e95c53c.

File: cheatsheets/Cryptographic_Storage_Cheat_Sheet.md [hunk 7]
```diff
@@ -1,234 +1,185 @@
...
 
-To perform the strong key management functions we have seen in requirement 3.6 we must have highly trusted and trained key custodians who understand how to perform key management duties. The key custodians must also sign a form stating they understand the responsibilities that come with this role.
+### Encrypting Stored Keys
 
-# Related documentation & tools
+Where possible, encryption keys should themselves be stored in an encrypted form. At least two separate keys are required for this:
 
-## Documentation
+- The Data Encryption Key (DEK) is used to encrypt the data.
+- The Key Encryption Key (KEK) is used to encryption the DEK.
```
Comment (Chick3nman) [inline cheatsheets/Cryptographic_Storage_Cheat_Sheet.md]: Change 'encryption' to 'encrypt'. Small grammatical change.

Comment (rbsec) [reply cheatsheets/Cryptographic_Storage_Cheat_Sheet.md]: Fixed

File: cheatsheets/Cryptographic_Storage_Cheat_Sheet.md (hunk 6 above)
Comment (rbsec) [reply cheatsheets/Cryptographic_Storage_Cheat_Sheet.md]: Good call - this was from the original CS and has some issues. I've just removed the algorithms in 220d6ee - the message is just as clear but without the list of (somewhat incorrect) examples cluttering it up.

**Comment by rbsec** [comment]
@ThunderSon @mackowski I've gone through for a grammar check, and I believe that this is now ready to merge. I've also updated the cross-references to it from a couple of other cheat sheets.

**Pull request merged**